tempesta_fw/t/unit/tfw_fuzzer.c

```diff
@@ -0,0 +1,563 @@
+				   gen_vector[HOST].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_x_forwarded_for(char *s1)
+{
+	struct fuzz_msg r = hosts[gen_vector[X_FORWARDED_FOR].i];
+
+	strcat(s1, "X-Forwarded-For:");
+	add_spaces(s1);
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[X_FORWARDED_FOR].i,
+				   gen_vector[X_FORWARDED_FOR].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[X_FORWARDED_FOR].i,
+				   gen_vector[X_FORWARDED_FOR].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_content_type(char *s1)
+{
+	struct fuzz_msg r = content_type[gen_vector[CONTENT_TYPE].i];
+
+	strcat(s1, "Content-Type:");
+	add_spaces(s1);
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, gen_vector[CONTENT_TYPE].i,
+				   gen_vector[CONTENT_TYPE].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, gen_vector[CONTENT_TYPE].i,
+				   gen_vector[CONTENT_TYPE].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_content_length(char *s1)
+{
+	struct fuzz_msg r = content_length[gen_vector[CONTENT_LENGTH].i];
+
+	strcat(s1, "Content-Length:");
+	add_spaces(s1);
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else
+		add_string(s1, gen_vector[CONTENT_LENGTH].i, NULL);
+
+	return FUZZ_INVALID;
+}
+
+static int __add_transfer_encoding(char *s1, int n)
+{
+	struct fuzz_msg r = transfer_encoding[n];
+
+	if (r.s) {
+		strcat(s1, r.s);
+		return r.inval;
+	} else {
+		if (gen_vector[CONTENT_LENGTH].i % 2) {
+			add_string(s1, n, gen_vector[n].a_val);
+			return FUZZ_VALID;
+		} else
+			add_string(s1, n, gen_vector[n].a_inval);
+	}
+
+	return FUZZ_INVALID;
+}
+
+static int add_transfer_encoding(char *s1)
+{
+	int v = 0, i;
+
+	strcat(s1, "Transfer-Encoding:");
+	add_spaces(s1);
+	v |= __add_transfer_encoding(s1, gen_vector[TRANSFER_ENCODING].i);
+	for (i = 0; i < gen_vector[TRANSFER_ENCODING_NUM].i; ++i) {
+		addch(s1, ',');
+		add_spaces(s1);
+		v |= __add_transfer_encoding(s1,
+				(i + gen_vector[CONTENT_LENGTH].i) %
+				gen_vector[TRANSFER_ENCODING].max);
+	}
+
+	return v;
+}
+
+static int add_body(char *s1, int type)
+{
+	size_t len = strlen(s1), cont_len,
+	       i, body_size = gen_vector[BODY_SIZE].i * 256;
+	char *rc;
+	int err;
+
+	for (i = 0; i < body_size; ++i)
+		s1[len + i] = a_body[((i + gen_vector[CONTENT_LENGTH].i +
+				       gen_vector[BODY_SIZE].i + len) ^
+				       (a_body[i] + len)) % sizeof(a_body)];
+
+	err = kstrtoul(content_length[gen_vector[CONTENT_LENGTH].i].s,
+			10, &cont_len);
+	if (err)
+		return FUZZ_INVALID;
+	if (cont_len > body_size)
+		return FUZZ_INVALID;
+	if (type == FUZZ_REQ)
+		return FUZZ_VALID;
+
+	rc = resp_code[gen_vector[RESP_CODE].i].s;
+	if (!strstr(rc, "100") || !strstr(rc, "101") || !strstr(rc, "102") ||
+			!strstr(rc, "204") || !strstr(rc, "304"))
+		return FUZZ_INVALID;
+
+	return FUZZ_VALID;
+}
```

Comment:
Chunked body should be tested, see https://tools.ietf.org/html/rfc7230#section-4.1